File: prep/implementation/patterns_reference_implementation/two_pointers/TripletsWithSmallerSumList.java
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;

// O(n3) / O(1)
public class TripletsWithSmallerSumList {
  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int tt = in.nextInt();

      while (tt-- > 0) {
        int n = in.nextInt();
        int s = in.nextInt();

        int[] a = new int[n];
        for (int i = 0; i < n; i++) {
          a[i] = in.nextInt();
        }

        Arrays.sort(a);

        List<List<Integer>> res = new ArrayList<>();
        for (int i = 0; i < n - 2; i++) {
          int left = i + 1, right = n - 1;

          while (left < right) {
            int sum = a[i] + a[left] + a[right];

            if (sum < s) {
              List<Integer> tmp = new LinkedList<>();
              for (int j = right; j > left; j--) {
                res.add(Arrays.asList(a[i], a[left], a[j]));
              }
              left++;
            } else {
              right--;
            }
          }
        }

        for (var sub : res) {
          for (int e : sub) {
            System.out.printf("%d ", e);
          }
          System.out.println();
        }
      }
    }
  }
}

File: prep/implementation/practice_iterations/two_pointers_2/triplets_with_smaller_sum_list.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

// O(n3) / O(n)
int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  int tt, n, s;
  cin >> tt;

  while (tt--) {
    cin >> n >> s;

    vector<int> a(n);
    for (int i = 0; i < n; i++) {
      cin >> a[i];
    }

    sort(a.begin(), a.end());

    vector<vector<int>> res;
    for (int i = 0; i < n - 2; i++) {
      int left = i + 1, right = n - 1;

      while (left < right) {
        int sum = a[i] + a[left] + a[right];

        if (sum < s) {
          for (int j = right; j > left; j--) {
            vector<int> sub;
            sub.push_back(a[i]);
            sub.push_back(a[left]);
            sub.push_back(a[j]);

            res.emplace_back(sub);
          }
          left++;
        } else {
          right--;
        }
      }
    }

    for (auto sub : res) {
      for (int e : sub) {
        cout << e << " ";
      }
      cout << "\n";
    }
  }

  return 0;
}